code/modules/mob/living/carbon/xenomorph/abilities.dm

```diff
@@ -287,13 +289,24 @@
 	update_button_icon()
 
 /datum/action/xeno_action/activable/secrete_resin/use_ability(atom/A)
+	//xeno owner of this ability
 	var/mob/living/carbon/xenomorph/xowner = owner
-	if(SSmonitor.gamestate == SHUTTERS_CLOSED && CHECK_BITFIELD(SSticker.mode.flags_round_type, MODE_ALLOW_XENO_QUICKBUILD) && SSresinshaping.active)
-		preshutter_build_resin(A)
-	else if(get_dist(owner, A) > xowner.xeno_caste.resin_max_range) //Maximum range is defined in the castedatum with resin_max_range, defaults to 0
-		build_resin(get_turf(owner))
+	var/turf/selectedturf = get_turf(xowner)
+	if(SSmonitor.gamestate == SHUTTERS_CLOSED)
+		if(selectedturf.get_build_prereqs())
+			if(get_dist(owner, A) > xowner.xeno_caste.resin_max_range) //Maximum range is defined in the castedatum with resin_max_range, defaults to 0
+				build_resin(get_turf(owner))
+			else
+				build_resin(get_turf(A))
+			return
+		else if(CHECK_BITFIELD(SSticker.mode.flags_round_type, MODE_ALLOW_XENO_QUICKBUILD) && SSresinshaping.active)
+			preshutter_build_resin(A)
 	else
-		build_resin(get_turf(A))
+		if(get_dist(owner, A) > xowner.xeno_caste.resin_max_range)
+			build_resin(get_turf(owner))
+		else
+			build_resin(get_turf(A))
+		return
```

Comment:
I can't suggest a change because there's a deleted line in the way, but I'm pretty sure this can be written simply as follows:
```
	var/turf/selectedturf = get_turf(xowner)
	if(SSmonitor.gamestate == SHUTTERS_CLOSED && CHECK_BITFIELD(SSticker.mode.flags_round_type, MODE_ALLOW_XENO_QUICKBUILD) && SSresinshaping.active && !selectedturf.get_build_prereqs())
		preshutter_build_resin(A)
		return
	
	if(get_dist(owner, A) > xowner.xeno_caste.resin_max_range)
		build_resin(get_turf(owner))
	else
		build_resin(get_turf(A))
```

I would test it to make sure though.